In [53]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix, classification_report

In [54]:
data = load_breast_cancer()
print(type(data))
print(dir(data))

<class 'sklearn.utils._bunch.Bunch'>
['DESCR', 'data', 'data_module', 'feature_names', 'filename', 'frame', 'target', 'target_names']


In [55]:
X = data.data
#print(X)

Y = data.target
#print(Y)

print(X.shape)

print(len(Y))

(569, 30)
569


In [56]:
#데이터 분할 (학슴 80%, 테스트 20%)
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42)
print(X_train.shape, Y_train.shape)
print(X_test.shape, Y_test.shape)

(455, 30) (455,)
(114, 30) (114,)


In [57]:
import platform
import sklearn 
print(platform.system())
if platform.system() == 'Windows':
    sklearn.set_config(display='text')

#모델 학습 (로지스틱 회귀)
model = LogisticRegression(max_iter=10000, random_state=42)
model.fit(X_train, Y_train)
print(dir(model))

Windows
['C', '__annotations__', '__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getstate__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__setstate__', '__sizeof__', '__sklearn_clone__', '__sklearn_tags__', '__str__', '__subclasshook__', '__weakref__', '_doc_link_module', '_doc_link_template', '_doc_link_url_param_generator', '_get_class_level_metadata_request_values', '_get_doc_link', '_get_fitted_attr_html', '_get_metadata_request', '_get_param_names', '_get_params_html', '_html_repr', '_init_callback_context', '_parameter_constraints', '_predict_proba_lr', '_repr_html_inner', '_repr_mimebundle_', '_validate_params', 'class_weight', 'classes_', 'coef_', 'decision_function', 'densify', 'dual', 'fit', 'fit_intercept', 'get_metadata_routing', 'get_params', 'intercept_', 'intercept_scaling', 'l

In [58]:
%who

BaseEstimator	 LogisticRegression	 MyFakeClassifier	 X	 X_test	 X_train	 Y	 Y_test	 Y_train	 
acc	 accuracy_score	 cf	 classification_report	 cm	 confusion_matrix	 data	 digits	 f1	 
f1_score	 fake_clf	 fake_pred	 load_breast_cancer	 load_digits	 model	 np	 pd	 platform	 
plt	 prec	 precision_score	 pred	 rec	 recall_score	 sklearn	 train_test_split	 y	 
y_pred	 y_prob	 y_test	 y_train	 


In [59]:
y_pred = model.predict(X_test)
print(y_pred)
#print(y_pred == y_test)
print(pd.Series(y_pred==Y_test).unique())
print(pd.Series(y_pred==Y_test).value_counts())
y_prob = model.predict_proba(X_test)
print(y_prob)

[1 0 0 1 1 0 0 0 1 1 1 0 1 0 1 0 1 1 1 0 1 1 0 1 1 1 1 1 1 0 1 1 1 1 1 1 0
 1 0 1 1 0 1 1 1 1 1 1 1 1 0 0 1 1 1 1 1 0 1 1 1 0 0 1 1 1 0 0 1 1 0 0 1 0
 1 1 1 1 1 1 0 1 1 0 0 0 0 0 1 1 1 1 1 1 1 1 0 0 1 0 0 1 0 0 1 1 1 0 1 1 0
 1 0 0]
[ True False]
True     109
False      5
Name: count, dtype: int64
[[1.21238266e-01 8.78761734e-01]
 [9.99999970e-01 3.03624981e-08]
 [9.98405090e-01 1.59490988e-03]
 [1.21353825e-03 9.98786462e-01]
 [1.39796565e-04 9.99860203e-01]
 [1.00000000e+00 1.94458720e-10]
 [1.00000000e+00 6.09483521e-11]
 [9.80773102e-01 1.92268979e-02]
 [1.58520339e-02 9.84147966e-01]
 [5.63029459e-03 9.94369705e-01]
 [7.04321239e-02 9.29567876e-01]
 [9.99193285e-01 8.06714616e-04]
 [1.10171295e-02 9.88982871e-01]
 [8.24174930e-01 1.75825070e-01]
 [7.79253365e-04 9.99220747e-01]
 [9.98272428e-01 1.72757217e-03]
 [1.18193344e-03 9.98818067e-01]
 [1.18481436e-05 9.99988152e-01]
 [5.99658938e-07 9.99999400e-01]
 [9.99999288e-01 7.11997904e-07]
 [1.71640882e-01 8.28359118e-01]
 [7.6396

In [60]:
cm = confusion_matrix(Y_test, y_pred)
print(cm)

[[39  4]
 [ 1 70]]


In [61]:
# 평가지표 계산 
acc = accuracy_score(Y_test, y_pred) #(정답, 예측)
pred = precision_score(Y_test, y_pred)  #positive에서 진짜 poisitive
rec = recall_score(Y_test, y_pred) 
f1 = f1_score(Y_test, y_pred)
print('평가지표')
print(f'Accuracy::{acc}')
print(f'Precision::{pred}')
print(f'Recall:{rec}')
print(f'F1 Score::{f1}')

평가지표
Accuracy::0.956140350877193
Precision::0.9459459459459459
Recall:0.9859154929577465
F1 Score::0.9655172413793104


#데이터가 불균형일 떄 발생되는 문제

불균형한 이진 분류로 만들기 위해 숫자 '7'만 True 나머지 숫자는 False로 설정

- 마찬가지로 fit은 아무것도 하지 않음
- predict에서는 무조건 0을 반환함
- 모든 예측을 다 0으로 하는 것! (모두 7이 아닌 다른 숫자라고 예측)
- 당연히 모델로서 말이 안 되는데 이것의 정확도를 측정

In [62]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.base import BaseEstimator
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
import numpy as np
import pandas as pd

In [63]:
class MyFakeClassifier(BaseEstimator):
    def fit(self,X,y):
        pass

    # 입력값으로 들어오는 X 데이터 셋의 크기만큼 모두 0값으로 만들어서 반환
    def predict(self,X):
        return np.zeros( (len(X), 1) , dtype=bool)

digits = load_digits()
print(digits.data)
print(digits.data.shape)

[[ 0.  0.  5. ...  0.  0.  0.]
 [ 0.  0.  0. ... 10.  0.  0.]
 [ 0.  0.  0. ... 16.  9.  0.]
 ...
 [ 0.  0.  1. ...  6.  0.  0.]
 [ 0.  0.  2. ... 12.  0.  0.]
 [ 0.  0. 10. ... 12.  1.  0.]]
(1797, 64)


In [64]:
print(digits.target)
y = (digits.target == 7).astype(int)
print(len(y))

print(pd.Series(y).value_counts())

X_train, X_test, y_train, y_test = train_test_split(digits.data, y, random_state=11)
print(pd.Series(y_train).value_counts())

print(pd.Series(y_test).value_counts())

[0 1 2 ... 8 9 8]
1797
0    1618
1     179
Name: count, dtype: int64
0    1213
1     134
Name: count, dtype: int64
0    405
1     45
Name: count, dtype: int64


In [65]:
fake_clf = MyFakeClassifier()
fake_clf.fit(X_train, y_train)

fake_pred = fake_clf.predict(X_test)
print('모든 예측을 0으로 하여도 정확도는 : {:.3f}'.format(accuracy_score(y_test, fake_pred)))

cf = confusion_matrix(y_test, fake_pred)
print(cf)

f1 = f1_score(y_test, fake_pred)
print(f'f1_score : {f1:.3f}')

prec = precision_score(y_test, fake_pred)
print(f'precision_score : {prec:.3f}')

모든 예측을 0으로 하여도 정확도는 : 0.900
[[405   0]
 [ 45   0]]
f1_score : 0.000
precision_score : 0.000


c:\sk-encore\data-collection-workspace\.venv\Lib\site-packages\sklearn\metrics\_classification.py:1879: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])


- 모든 예측을 0으로 했지만, 0 (7이 아닌 숫자)이 1 (숫자 7)보다 훨씬 많은 불균형 데이터였기 때문에, 정확도가 무려 90%가 나왔다.

- 이렇게 두 가지 예제로 불균형 데이터에서 정확도의 성능 평가지표로서 한계를 보였다

In [69]:
print("average=None (각 클래스별로 계산)")
print("Precision:", precision_score(y_test, fake_pred, average=None))
print("Recall:", recall_score(y_test, fake_pred, average=None))
print("F1:", f1_score(y_test, fake_pred, average=None))
print()

average=None (각 클래스별로 계산)
Precision: [0.9 0. ]
Recall: [1. 0.]
F1: [0.94736842 0.        ]



c:\sk-encore\data-collection-workspace\.venv\Lib\site-packages\sklearn\metrics\_classification.py:1879: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])


0. average='binary' (default)
   - positive label을 1로 간주하고 계산한다.
   - y_true에 0과 1 두 개 클래스만 있어야 한다.

1. average=None (클래스별 성능 개별 출력)

| 클래스               | Precision | Recall | F1-score | Support |
| ----------------- | --------- | ------ | -------- | ------- |
| **0 (malignant)** | 0.97      | 0.91   | 0.94     | 43      |
| **1 (benign)**    | 0.95      | 0.99   | 0.97     | 71      |

- 각 클래스별로 정밀도, 재현율, F1-score를 개별적으로 계산하여 보여준다.
- 모델이 클래스 0 (악성 종양)을 상대적으로 덜 잘 맞추고 있다는 것을 파악할 수 있다.


In [72]:
y_pred = model.predict(X_test)   # 반드시 X_test로 예측!

print(len(y_test))   # 114
print(len(y_pred))    # 114 (같아야 함)

precision_score(y_test, y_pred, average='micro')

ValueError: X has 64 features, but LogisticRegression is expecting 30 features as input.

In [73]:
print(X_test.shape)   # (샘플수, 특성수) 형태로 나옴
print(X_train.shape)

(450, 64)
(1347, 64)


In [74]:
model = LogisticRegression(max_iter=10000, random_state=42)
model.fit(X_train, y_train)   # ← 이 셀을 반드시 다시 실행!

y_pred = model.predict(X_test)

In [75]:
print(model.n_features_in_)   # 지금 model이 몇 개 특성으로 학습됐는지 확인

64


In [76]:
print("model이 학습한 특성 수:", model.n_features_in_)
print("X_test의 특성 수:", X_test.shape[1])

y_pred = model.predict(X_test)
print("예측 완료!")
print(y_pred)

model이 학습한 특성 수: 64
X_test의 특성 수: 64
예측 완료!
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 0 0 0 0 0 1 0
 0 0 0 1 1 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 1 0 1 0 0
 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 0 0 1 0 0 0 0 0 0 0 0 0 0 1 0 0 1 1 0
 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 1 0 0 0 0 0 0 0 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 1 0 0 0 1 0 0 0 0 0 0 0 0 0 0 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 1 0 0 0 1 0 0 0 1 0 0 0 1 0 0 0 0 0 0 1 0 0 0 1 0 1 0 0 0 0 0 0 0 0 0
 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1 0 1 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 1 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0]


In [78]:
from sklearn.metrics import classification_report


target_names = data.target_names  # ['malignant', 'benign']
print(target_names)

'''
| 용어            | 한글 해석   | 설명                                                                          |
| -------------   | ------- | --------------------------------------------------------------------------- |
| **malignant**   | 악성 (암)  | 빠르게 자라고 전이할 가능성이 높은 종양이나 질병. 보통 **암(Cancer)** 을 뜻함. 치료가 어렵고 생명에 위협이 될 수 있음. |
| **benign**      | 양성  (비암성) | 성장하긴 하지만 전이하지 않으며 비교적 **위험하지 않은 상태**. 수술 등으로 완치 가능한 경우가 많음.                 |

- malignant: 악성 종양 (암) → 0 또는 Negative Class
- benign: 양성 종양 (비암성) → 1 또는 Positive Class
'''



# 0. 전체 classification_report
print("Classification Report")
# target_names옵션을 설정하지 않으면 0과 1로 표시된다.
print(classification_report(y_test, y_pred))
print("Classification Report")
print(classification_report(y_test, y_pred, target_names=target_names))

['malignant' 'benign']
Classification Report
              precision    recall  f1-score   support

           0       1.00      0.99      0.99       405
           1       0.91      0.96      0.93        45

    accuracy                           0.99       450
   macro avg       0.95      0.97      0.96       450
weighted avg       0.99      0.99      0.99       450

Classification Report
              precision    recall  f1-score   support

   malignant       1.00      0.99      0.99       405
      benign       0.91      0.96      0.93        45

    accuracy                           0.99       450
   macro avg       0.95      0.97      0.96       450
weighted avg       0.99      0.99      0.99       450



In [77]:
# 2. average='micro'
print("average='micro'")
print("Precision:", precision_score(y_test, y_pred, average='micro'))
print("Recall   :", recall_score(y_test, y_pred, average='micro'))
print("F1 Score :", f1_score(y_test, y_pred, average='micro'))
print()

# 3. average='macro'
print("average='macro'")
print("Precision:", precision_score(y_test, y_pred, average='macro'))
print("Recall   :", recall_score(y_test, y_pred, average='macro'))
print("F1 Score :", f1_score(y_test, y_pred, average='macro'))
print()

# 4. average='weighted'
print("average='weighted'")
print("Precision:", precision_score(y_test, y_pred, average='weighted'))
print("Recall   :", recall_score(y_test, y_pred, average='weighted'))
print("F1 Score :", f1_score(y_test, y_pred, average='weighted'))
print()

average='micro'
Precision: 0.9866666666666667
Recall   : 0.9866666666666667
F1 Score : 0.9866666666666667

average='macro'
Precision: 0.9549654189324746
Recall   : 0.9728395061728395
F1 Score : 0.9636784330606973

average='weighted'
Precision: 0.9870228604614328
Recall   : 0.9866666666666667
F1 Score : 0.9867950925527337



2. average='micro' (전체 기준 - 샘플 단위의 합)  
- micro 평균은 모든 TP, FP, FN을 전 클래스 기준으로 합산한 뒤 계산한다.
- 이 경우 정확도(Accuracy) = Precision = Recall = F1-score가 된다.
- 총 정답 수: TP(전체 맞춤 수) = 0 클래스에서 39개 + 1 클래스에서 70개 = 109
- 전체 샘플 수: 114
- 따라서:    Accuracy = Precision = Recall = F1 = 109 / 114 ≈ 0.9561  

3. average='macro' (단순 평균)  
- 클래스별 점수를 동일한 비중으로 평균함
  Macro Precision = (0.97 + 0.95) / 2 = 0.96  
  Macro Recall    = (0.91 + 0.99) / 2 = 0.95  
  Macro F1-score  = (0.94 + 0.97) / 2 = 0.955 ≈ 0.95  

- 클래스 0과 1의 비중이 같다고 가정하고 평균 내기 때문에, 클래스 불균형이 있는 경우에는 왜곡이 있을 수 있다.

4. average='weighted' (샘플 수 기반 가중 평균)  
-  각 클래스의 점수에 해당 클래스의 샘플 개수(Support)를 곱해서 평균
  Weighted Precision = (0.97×43 + 0.95×71) / 114 ≈ 0.958    
  Weighted Recall    = (0.91×43 + 0.99×71) / 114 ≈ 0.961  
  Weighted F1-score  = (0.94×43 + 0.97×71) / 114 ≈ 0.961  

- 실제 출력값은 모두 반올림되어 0.96으로 나옴

- 데이터 불균형을 고려한 평균이므로, 실전에서 가장 신뢰도 있는 지표 중 하나임

5. average 옵션 요약

| 평균 방식      | 설명                    | 특징               | 언제 사용하면 좋은가?                 |
| ---------- | --------------------- |------------ | ---------------------------- |
| `None`     | 클래스별 개별 지표 확인         |  `array` 형태       | 특정 클래스 성능 비교 시               |
| `micro`    | 전체 샘플 기준으로 정밀도/재현율 계산 | 불균형 데이터에 취약      | 다중 클래스, 불균형 무시하고 전체 성능 확인 시  |
| `macro`    | 클래스별 단순 평균            | 클래스 중요도 균일할 때 사용 |클래스 간 균형 평가 시                |
| `weighted` | 클래스별 가중 평균            | 불균형 데이터에 적합      |**불균형 데이터에서의 전체 성능 평가** 시 유리 |


# 1. LogisticRegression  분류 알고리즘

LogisticRegression( )알고리즘의 penalty='l2', *, dual=False, tol=0.0001, C=1.0, fit_intercept=True, intercept_scaling=1, class_weight=None, random_state=None, solver='lbfgs', max_iter=100, multi_class='deprecated', verbose=0, warm_start=False, n_jobs=None, l1_ratio=None  


1. **penalty:**  
정규화 방법을 지정한다. 'l2'는 L2 정규화를 의미하며, 이는 가중치의 제곱합을 최소화하여 과적합을 방지하는 데 도움을 준다. 'l1'을 선택하면 L1 정규화를 사용하여 일부 가중치를 0으로 만들 수 있다. 'elasticnet'을 선택하면 L1과 L2 정규화를 혼합하여 사용할 수 있다.  
**2. dual:**  
이진 분류 문제에서 이중 문제를 해결할지를 결정한다. True로 설정하면 이중 문제를 해결하고, False로 설정하면 원래의 문제를 해결한다. 일반적으로 이진 분류에서는 False로 설정하는 것이 더 효율적이다.  
**3. tol:**  
수렴 기준을 설정합니다. 알고리즘이 수렴했다고 판단하는 기준으로, 기본값은 0.0001이다. 이 값이 작을수록 더 정밀한 수렴을 요구한다.  
**4. C:**  
정규화 강도를 조절하는 하이퍼파라미터이다. 값이 작을수록 정규화가 강해지고, 값이 클수록 정규화가 약해진다. 기본값은 1.0이다.  
**5. fit_intercept:**  
절편(intercept)을 모델에 포함할지를 결정한다. True로 설정하면 절편을 포함하고, False로 설정하면 절편을 포함하지 않는다. 기본값은 True이다.  
**6. intercept_scaling:**  
다중 클래스 문제에서 절편을 조정하는 데 사용된다. 기본값은 1이며, 이 값은 절편의 스케일을 조정하는 데 사용된다.  
**7. class_weight:**  
클래스의 가중치를 설정한다. None으로 설정하면 모든 클래스의 가중치가 동일하게 설정됩니다. 클래스 불균형이 있는 경우, 특정 클래스에 더 높은 가중치를 부여하여 모델이 해당 클래스를 더 잘 학습하도록 할 수 있다.
**8. random_state:**  
난수 생성기의 시드를 설정합니다. 동일한 시드를 사용하면 동일한 결과를 재현할 수 있다. 기본값은 None이다.  
**9. solver:**  
최적화 알고리즘을 선택한다. 'lbfgs', 'newton-cg', 'liblinear', 'sag', 'saga' 중 하나를 선택할 수 있다. 각 솔버는 서로 다른 특성과 성능을 가지므로 데이터셋에 따라 적절한 솔버를 선택하는 것이 중요하다.  
**10. max_iter:**      
알고리즘의 최대 반복 횟수를 설정한다. 기본값은 100이며, 이 값에 도달하기 전에 수렴하면 알고리즘이 종료된다.  
**11. multi_class:**  
다중 클래스 문제에서의 처리 방법을 설정한다. 'deprecated'는 더 이상 사용되지 않는 옵션이며, 대신 'ovr' (One-vs-Rest) 또는 'multinomial'을 사용할 수 있다.  
**12. verbose:**  
출력의 상세 정도를 설정한다. 0이면 출력이 없고, 1 이상의 값으로 설정하면 더 많은 정보가 출력된다.  
**13. warm_start:**  
이전에 학습한 모델을 재사용할지를 결정한다. True로 설정하면 이전 모델의 가중치를 초기값으로 사용하여 학습을 계속할 수 있다.  
**14. n_jobs:**  
병렬 처리를 위한 작업 수를 설정한다. None으로 설정하면 기본값이 사용되며, -1로 설정하면 모든 CPU 코어를 사용한다.  
**15. l1_ratio:**  
L1과 L2 정규화의 비율을 설정한다. 'elasticnet'을 사용할 때만 의미가 있으며, 기본값은 None이다.